### PR TITLE
Ensemble builder

### PR DESCRIPTION
**Pull request by franchuterivera**
This is a proof of concept PR.  It is fully functional, yet it assumes we cannot merge the metrics from Auto-Sklearn and Auto-Pytorch.

Today they differ slightly and for those reasons, I have proposed this or to take in a loss function when doing ensemble selection.

It also removed many lines that are legacy.

### CODE REVIEW

File: common/ensemble_building/singlebest_ensemble.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+        self.best_loss = best_model_loss
+
+        return best_model_identifier
+
+    def predict(self, predictions: Union[np.ndarray, List[np.ndarray]]) -> np.ndarray:
+        return predictions[0]
+
+    def __str__(self) -> str:
+        return "Single Model Selection:\n\tMembers: %s" "\n\tWeights: %s\n\tIdentifiers: %s" % (
```
Comment (nabenabe0928) [inline common/ensemble_building/singlebest_ensemble.py]: Can we avoid using %string format? (because it was not recommended officially)
We can use `.format()` or f-string instead.
In terms of the speed itself, f-string is better, but since this process would not dominate the whole implementation, `format()` is a better option.
In any case, %string is slower than f-string and messier than `.format`, so we will not be benefited from %string.

File: common/ensemble_building/singlebest_ensemble.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    def fit(
+        self,
+        base_models_predictions: np.ndarray,
+        true_targets: np.ndarray,
+        model_identifiers: List[Tuple[int, int, float]],
+    ) -> "AbstractEnsemble":
+        return self
+
+    def get_identifiers_from_run_history(self) -> List[Tuple[int, int, float]]:
```
Comment (nabenabe0928) [inline common/ensemble_building/singlebest_ensemble.py:48]: We can write the code more concisely as follows:
```
min_idx = np.array([run_value.cost for run_value in self.run_history.data.values()]).argmin()
min_key = list(self.run_history.data.keys())[min_idx]
min_run_value = self.run_history.data[min_key]

if min_run_value < best_model_loss:
    ...
else:  # SMAC did not work properly
    raise ValueError
```

File: common/ensemble_building/singlebest_ensemble.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        seed: int,
+        backend: Backend,
+    ):
+        self.seed = seed
+        self.backend = backend
+
+        # Add some default values -- at least 1 model in ensemble is assumed
+        self.indices_ = [0]
+        self.weights_ = [1.0]
+        self.run_history = run_history
```
Comment (nabenabe0928) [inline common/ensemble_building/singlebest_ensemble.py:36]: Can we pass `run_history` to functions when we really need it? (To make this code more safe)

File: common/ensemble_building/singlebest_ensemble.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        self,
+        run_history: RunHistory,
+        seed: int,
+        backend: Backend,
+    ):
+        self.seed = seed
+        self.backend = backend
+
+        # Add some default values -- at least 1 model in ensemble is assumed
+        self.indices_ = [0]
```
Comment (nabenabe0928) [inline common/ensemble_building/singlebest_ensemble.py:34]: Why `indices_`, but not `_indices`?
`foo_` is for reserved words such as `class`, `list`.
`_foo` is for private variables.

File: common/ensemble_building/ensemble_selection.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+    def __init__(
+        self,
+        ensemble_size: int,
+        loss_fn: Callable[..., float],
+        loss_fn_args: Dict[str, Any],
+        random_state: np.random.RandomState,
+    ) -> None:
+        self.ensemble_size = ensemble_size
+        self.loss_fn: Optional[Callable[..., float]] = loss_fn
+        self.loss_fn_args = loss_fn_args
```
Comment (nabenabe0928) [inline common/ensemble_building/ensemble_selection.py:25]: These are private variables.

File: common/ensemble_building/ensemble_selection.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+    def fit(
+        self,
+        predictions: List[np.ndarray],
+        labels: np.ndarray,
+        identifiers: List[Tuple[int, int, float]],
+    ) -> AbstractEnsemble:
+        self.ensemble_size = int(self.ensemble_size)
+        if self.ensemble_size < 1:
+            raise ValueError("Ensemble size cannot be less than one!")
```
Comment (nabenabe0928) [inline common/ensemble_building/ensemble_selection.py:49]: I often see `foo must be something` rather than `foo cannot be something`

File: common/ensemble_building/ensemble_selection.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+        weighted_ensemble_prediction = np.zeros(
+            predictions[0].shape,
+            dtype=np.float64,
+        )
+        fant_ensemble_prediction = np.zeros(
+            weighted_ensemble_prediction.shape,
+            dtype=np.float64,
+        )
+        for _ in range(ensemble_size):
+            losses = np.zeros(
```
Comment (nabenabe0928) [inline common/ensemble_building/ensemble_selection.py]: `losses.fill(0.0, dtype=np.float64)` is faster.

File: common/ensemble_building/ensemble_selection.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+            # Memory-efficient averaging!
+            for j, pred in enumerate(predictions):
+                # fant_ensemble_prediction is the prediction of the current ensemble
+                # and should be ([predictions[selected_prev_iterations] + predictions[j])/(s+1)
+                # We overwrite the contents of fant_ensemble_prediction
+                # directly with weighted_ensemble_prediction + new_prediction and then scale for avg
+                np.add(weighted_ensemble_prediction, pred, out=fant_ensemble_prediction)
+                np.multiply(
+                    fant_ensemble_prediction, (1.0 / float(s + 1)), out=fant_ensemble_prediction
+                )
```
Comment (nabenabe0928) [inline common/ensemble_building/ensemble_selection.py:111]: This part consumes the same amount as the following:
```
fant_ensemble_prediction *= 1.0 / float(s + 1)
```
It is shorter and understandable

File: common/ensemble_building/ensemble_selection.py [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+                    out=weighted_ensemble_prediction,
+                )
+
+            # Memory-efficient averaging!
+            for j, pred in enumerate(predictions):
+                # fant_ensemble_prediction is the prediction of the current ensemble
+                # and should be ([predictions[selected_prev_iterations] + predictions[j])/(s+1)
+                # We overwrite the contents of fant_ensemble_prediction
+                # directly with weighted_ensemble_prediction + new_prediction and then scale for avg
+                np.add(weighted_ensemble_prediction, pred, out=fant_ensemble_prediction)
```
Comment (nabenabe0928) [inline common/ensemble_building/ensemble_selection.py:108]: Only this part (line 99) is memory-efficient and it is worth keeping the current form.

File: common/ensemble_building/ensemble_selection.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+            dtype=np.float64,
+        )
+        for _ in range(ensemble_size):
+            losses = np.zeros(
+                (len(predictions)),
+                dtype=np.float64,
+            )
+            s = len(ensemble)
+            if s > 0:
+                np.add(
```
Comment (nabenabe0928) [inline common/ensemble_building/ensemble_selection.py:96]: This part consumes the same amount as the following:
```
weighted_ensemble_prediction += ensemble[-1]
```
It is shorter and understandable.

Comment (nabenabe0928) [inline common/ensemble_building/ensemble_selection.py:96]: This part consumes the same amount as the following:
```
weighted_ensemble_prediction += ensemble[-1]
```

File: common/ensemble_building/ensemble_selection.py [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+
+        self.num_input_models_ = len(predictions)
+
+        ensemble: List[np.ndarray] = []
+        trajectory: List[float] = []
+        order: List[int] = []
+
+        ensemble_size = self.ensemble_size
+
+        weighted_ensemble_prediction = np.zeros(
```
Comment (nabenabe0928) [inline common/ensemble_building/ensemble_selection.py:77]: How about making the dtype `np.float32`?
I think the scale of loss function values is within 9 digits and then it makes sense to use `np.float32`.

File: common/ensemble_building/ensemble_selection.py [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+            if len(predictions) == 1:
+                break
+
+        self.indices_ = order
+        self.trajectory_ = trajectory
+        self.train_loss_ = trajectory[-1]
+        return self
+
+    def _calculate_weights(self) -> None:
+        ensemble_members = Counter(self.indices_).most_common()
```
Comment (nabenabe0928) [inline common/ensemble_building/ensemble_selection.py:131]:
```
# List[Tuple[int, int]]: List[(index of model, occurrences)]
```

File: common/ensemble_building/ensemble_selection.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+        if len(predictions) == len(self.weights_):
+            for pred, weight in zip(predictions, self.weights_):
+                np.multiply(pred, weight, out=tmp_predictions)
+                np.add(average, tmp_predictions, out=average)
+
+        # if prediction model.shape[0] == len(non_null_weights),
+        # predictions do not include those of zero-weight models.
+        elif len(predictions) == np.count_nonzero(self.weights_):
+            non_null_weights = [w for w in self.weights_ if w > 0]
+            for pred, weight in zip(predictions, non_null_weights):
```
Comment (nabenabe0928) [inline common/ensemble_building/ensemble_selection.py:162]: `non_zero_weights`?

File: common/ensemble_building/singlebest_ensemble.py (hunk 1 above)
Comment (franchuterivera) [reply common/ensemble_building/singlebest_ensemble.py]: Thanks, I have changed any log/msg with this.

File: common/ensemble_building/singlebest_ensemble.py (hunk 2 above)
Comment (franchuterivera) [reply common/ensemble_building/singlebest_ensemble.py:48]: But there are plenty of this to check, like if the path exists (it might have been deleted by ensemble builder), among others. This suggestion does not align to that. Yes, one could do list comprehension with filtering, but then it is not much different than the actual code...

File: common/ensemble_building/singlebest_ensemble.py (hunk 3 above)
Comment (franchuterivera) [reply common/ensemble_building/singlebest_ensemble.py:36]: The reason I save it to the object is so that it is stored on it for debug. If you just pass it around and need this small dict like object, then you won't have this info.

I prefer to keep it if it is ok.

File: common/ensemble_building/singlebest_ensemble.py (hunk 4 above)
Comment (franchuterivera) [reply common/ensemble_building/singlebest_ensemble.py:34]: But this is a private variable?

File: common/ensemble_building/ensemble_selection.py (hunk 7 above)
Comment (franchuterivera) [reply common/ensemble_building/ensemble_selection.py]: Actually, this line is not even needed, I have moved it outside of the for loop.

File: common/ensemble_building/ensemble_selection.py (hunk 8 above)
Comment (franchuterivera) [reply common/ensemble_building/ensemble_selection.py:111]: I would prefer to leave the `out=` so people realize that we want to use the same array.

Yes `+=` is equal but `a = a + b` isn't, and the latter could cause memory peaks in the future. So it is better to have it more explicit.

File: common/ensemble_building/ensemble_selection.py (hunk 11 above)
Comment (franchuterivera) [reply common/ensemble_building/ensemble_selection.py:77]: I have added a precision attribute for any user to control the precision. 

The users of this class can control as needed.

File: common/ensemble_building/ensemble_selection.py (hunk 13 above)
Comment (franchuterivera) [reply common/ensemble_building/ensemble_selection.py:162]: Thanks, I have enhanced the message.

File: common/ensemble_building/ensemble_selection.py (hunk 6 above)
Comment (franchuterivera) [reply common/ensemble_building/ensemble_selection.py:49]: Hi, do you prefer this to be changed? If so, can you propose a github suggestion?

File: common/ensemble_building/ensemble_selection.py (hunk 5 above)
Comment (franchuterivera) [reply common/ensemble_building/ensemble_selection.py:25]: This can be an estimator at the end, so we have to have the same name in the initialization arguments as the attributes.

File: common/ensemble_building/singlebest_ensemble.py (hunk 4 above)
Comment (nabenabe0928) [reply common/ensemble_building/singlebest_ensemble.py:34]: I do know the usage of `indices`, so I'm partially asking.
If you use `indices_`, you should specify the intentions.

File: common/ensemble_building/singlebest_ensemble.py (hunk 2 above)
Comment (nabenabe0928) [reply common/ensemble_building/singlebest_ensemble.py:48]: I will check here later (my time schedule is temporally tight..., I think I can check on this weekend)
